certbot/plugins/common.py

```diff
@@ -99,7 +100,131 @@ def dest(self, var):
+        self.initialized = True
+
+    def load(self):
+        """Reads PluginStorage content from the disk to a dict structure
+
+        :returns: Plugin storage object
+        :rtype: dict
+
+        :raises .errors.PluginStorageError: when unable to open or read the file
+        """
+        data = dict()
+        filedata = ""
+        try:
+            with open(self.storagepath, 'r') as fh:
+                filedata = fh.read()
+        except IOError as e:
+            errmsg = "Could not read PluginStorage data file: {0} : {1}".format(
+                self.storagepath, str(e))
+            if os.path.isfile(self.storagepath):
+                # Only error out if file exists, but cannot be read
+                logger.error(errmsg)
+                raise errors.PluginStorageError(errmsg)
+        except TypeError:
+            # This happens if storagepath is not set, eg. None
+            raise errors.PluginStorageError("Could not open PluginStorage")
+        try:
+            data = json.loads(filedata)
+        except ValueError:
+            if len(filedata) == 0:
+                logger.debug("Plugin storage file %s was empty, no values loaded",
+                             self.storagepath)
+            else:
+                errmsg = "PluginStorage file {0} is corrupted.".format(
+                    self.storagepath)
+                logger.error(errmsg)
+                raise errors.PluginStorageError(errmsg)
+        return data
+
+    def save(self):
+        """Saves PluginStorage content to disk
+
+        :raises .errors.PluginStorageError: when unable to serialize the data
+            or write it to the filesystem
+        """
+        if not self.initialized or not self.storagepath:
+            errmsg = "Unable to save, problem with configuration directory"
+            logger.error(errmsg)
+            raise errors.PluginStorageError(errmsg)
+
+        try:
+            serialized = json.dumps(self._data)
+        except TypeError as e:
+            errmsg = "Could not serialize PluginStorage data: {0}".format(
+                str(e))
+            logger.error(errmsg)
+            raise errors.PluginStorageError(errmsg)
+        try:
+            with os.fdopen(os.open(self.storagepath,
+                                   os.O_WRONLY | os.O_CREAT, 0o600), 'w') as fh:
+                fh.truncate()
+                fh.write(serialized)
+        except IOError as e:
+            errmsg = "Could not write PluginStorage data to file {0} : {1}".format(
+                self.storagepath, str(e))
+            logger.error(errmsg)
+            raise errors.PluginStorageError(errmsg)
+
+    def put(self, key, value):
+        """Put configuration value to PluginStorage
+
+        :param str key: Key to store the value to
+        :param value: Data to store
+        """
+        if not self.initialized:
+            self.initialize_storage()
+
+        if not self.classkey in self._data.keys():
```

Comment:
If you wanted, we could make this one line like:
```
self._data.setdefault(self.classkey, dict())[key] = value
```

I don't have a strong preference though so if you like it better this way, keep it as is!

Reply:
Decided to keep it as it is.